### ISSUE TITLE
VolumeMounts not found: "datadog"

### ISSUE THREAD

**Issue by gurpalw** [DataDog/helm-charts#1041]
Previous issue here: https://github.com/DataDog/helm-charts/issues/986 closed without the chart being fixed. still present in 3.29

Describe what happened:
Upgrading the datadog agents from chart 3.10.6 to higher versions results in no agents being created and the daemonset rollout never progresses. I do not know which version introduces this behavior, but have confirmed that I can replicate the issue when going from 3.10.6 to the following versions:

    3.21.x
    3.22.x
    3.23.x
    3.25.x
    3.29.0

Related Events:

Error creating: Pod "datadog-96z27" is invalid: [spec.initContainers[0].volumeMounts[2].name: Not found: "datadog", spec.initContainers[1].volumeMounts[5].name: Not found: "datadog"]

Describe what you expected:

I expected the daemonset rollout to progress and complete

Steps to reproduce the issue:

Deploy chart 3.10.6
Upgrade to 3.25.1

Additional environment details (Operating System, Cloud provider, etc):
Chart: 3.29.0
Provider: EKS
K8s Version: 1.25.5

**Comment by iahmad-khan** [comment]
we can check all the 3xx versions, using helm template command, most of them , including the latest version is not creating DS yaml, and also the ones that create DS yaml, doesnt create an volumes for agent if we need to mount a volume on agent

**Comment by JacksonDavenport** [comment]
Hello, so this issue should be fixed with the near upcoming `7.45.0` release of the Cluster Agent. As a technical note the underlying issue itself is a bug in the Cluster Agent where when running the Admission Controller with `clusterAgent.admissionController.mutateUnlabelled: true`. What can happen is as this mutates all pods this includes the Datadog Agent pods, the Cluster Agent tries to avoid injecting the mounts when not necessary but got into the state where the `volumeMounts` for Unix-Domain-Socket (UDS) transport of APM data were injected into the `initContainers`, but the host `volume` that these were referencing were not injected. Causing this `datadog` volume not found error as the deployed Pod was invalid. 

This Cluster Agent bug technically has existed for awhile, but was exposed more in Helm Chart Version `3.22.0` where the Admission Controller was updated to use the same injection mode as your APM mode in the Helm config. Which since the APM Socket mode is enabled by default could have caused that switch in behavior. 

So for this we recommend to upgrade to the Cluster Agent `7.45.0` when available. However for workarounds in the meantime you can:

### Workaround Options
1. Change the Admission Controller Mutate Unlabelled back to false (the default)

Note: This will disable the Admission Controller’s injection attempt on the Agent pod, as it doesn’t have the label to trigger this by default. However, if you are relying on this injection for your other APM application pods you will need to add that injection label

```yaml
datadog:
  #(...)

clusterAgent:
  admissionController:
    enabled: true
    # Add Pod label admission.datadoghq.com/enabled: "true" to trigger
    mutateUnlabelled: false
```

2. Manually set the Admission Controller’s config mode to “hostip” or “service”, to avoid using the “socket“ mode. 

Note: these would require that the Agent has datadog.apm.portEnabled: true to receive those traces over that instead. But will allow the continued use of mutateUnlabelled: true. 

```yaml
datadog:
  #(...)
  apm:
    portEnabled: true
    
clusterAgent:
  admissionController:
    enabled: true
    mutateUnlabelled: true
    configMode: service
```

3. Update the Agent pods to explicitly disable admission controller

Note: This will allow you to keep the settings mutate unlabelled as true and injection mode as socket, just explicitly telling the Cluster Agent to avoid updating the Agent pods

```yaml
datadog:
  #(...)
  
agents:
  # Disable Admission Controller for Agent pods
  additionalLabels:
    admission.datadoghq.com/enabled: "false"
```


I'd probably recommend option 3 as its the most immediate in terms of the fix. Just to note as well if the pods are failing to deploy the Kubernetes DaemonSet controller will probably only re-attempt to deploy them every 15 minutes. So if using Option 1 or 2 you may see it take 15 minutes for the Agent pods to be re-deployed. If you want you can delete the Agent DaemonSet first and then `helm upgrade` to shortcut this process.

**Comment by JacksonDavenport** [comment]
Closing this issue with the release of 7.45.0 of the Cluster Agent which should fix the bug in the Cluster Agent that causes this state.
